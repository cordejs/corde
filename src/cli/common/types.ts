import { Command } from "commander";

export type ParamsFrom = "args" | "options" | "both";

export interface ICommandOptions {
  flags: string;
  description?: string;
  defaultValue?: string;
}

export interface ICliCommandConstructor {
  name?: string;
  program: Command;
  paramsFrom?: ParamsFrom;
}

export interface ICliCommand {
  readonly paramsFrom?: ParamsFrom;
  readonly command: Command;
  handler(...args: any[]): Promise<void> | void;
}

export interface ICliCommandConstructable {
  new (command: Command): ICliCommand;
}
