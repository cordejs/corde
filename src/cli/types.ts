export interface ICommand {
  action(...args: any[]): Promise<void> | void;
}
