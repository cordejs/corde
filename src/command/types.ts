import { ICordeBot } from "../types";

export interface ICreateMatcherParam {
  matcher?: string;
  isNot?: boolean;
  commandName?: string | boolean | number;
  channelId?: string;
  guildId?: string;
  isDebug?: boolean;
  isCascade?: boolean;
  cordeBot?: ICordeBot;
  trace?: string;
  utility?: boolean;
  mustSendCommand?: boolean;
}
