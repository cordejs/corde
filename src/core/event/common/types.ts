import { Client, ClientEvents } from "discord.js";

export type EventNames = keyof ClientEvents;

export type Spread<T extends EventNames> = ClientEvents[T][1] extends undefined
  ? ClientEvents[T] extends (infer U)[]
    ? U
    : never
  : ClientEvents[T];

export interface IDiscordEvent<TEvent extends EventNames, TFilter> {
  canListen(): boolean;
  getIntents(): string[];
  on(fn: (...args: ClientEvents[TEvent]) => void): this;
  once(filter?: TFilter): Promise<Spread<TEvent>>;
}

export interface IDiscordEventConstructable<TEvent extends EventNames, TFilter> {
  new (client: Client): IDiscordEvent<TEvent, TFilter>;
}

export type IntentType =
  | "GUILDS"
  | "GUILD_MEMBERS"
  | "GUILD_BANS"
  | "GUILD_EMOJIS_AND_STICKERS"
  | "GUILD_INTEGRATIONS"
  | "GUILD_WEBHOOKS"
  | "GUILD_INVITES"
  | "GUILD_VOICE_STATES"
  | "GUILD_PRESENCES"
  | "GUILD_MESSAGES"
  | "GUILD_MESSAGE_REACTIONS"
  | "GUILD_MESSAGE_TYPING"
  | "DIRECT_MESSAGES"
  | "DIRECT_MESSAGE_REACTIONS"
  | "DIRECT_MESSAGE_TYPING"
  | "GUILD_SCHEDULED_EVENTS";

export type EventClass = {
  [k in IntentType]: {
    bitfield: number;
    events: EventNames[];
  };
};
