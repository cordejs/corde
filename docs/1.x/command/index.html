<!doctype html>
<html lang="en" dir="ltr">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.0">
<link rel="alternate" type="application/rss+xml" href="/blog/rss.xml" title="Corde Blog RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/blog/atom.xml" title="Corde Blog Atom Feed"><title data-react-helmet="true">Command | Corde</title><meta data-react-helmet="true" property="og:url" content="https://cordejs.github.io/corde/docs/1.x/command"><meta data-react-helmet="true" name="docusaurus_locale" content="en"><meta data-react-helmet="true" name="docusaurus_version" content="1.x"><meta data-react-helmet="true" name="docusaurus_tag" content="docs-default-1.x"><meta data-react-helmet="true" property="og:title" content="Command | Corde"><meta data-react-helmet="true" name="description" content="When testing functionalities with an bot, expect will let you define what a bot should do"><meta data-react-helmet="true" property="og:description" content="When testing functionalities with an bot, expect will let you define what a bot should do"><link data-react-helmet="true" rel="shortcut icon" href="/img/logo/favicon.ico"><link data-react-helmet="true" rel="canonical" href="https://cordejs.github.io/corde/docs/1.x/command"><link data-react-helmet="true" rel="alternate" href="https://cordejs.github.io/corde/docs/1.x/command" hreflang="en"><link data-react-helmet="true" rel="alternate" href="https://cordejs.github.io/corde/docs/1.x/command" hreflang="x-default"><link rel="stylesheet" href="/assets/css/styles.e0889f48.css">
<link rel="preload" href="/assets/js/runtime~main.de5db1e2.js" as="script">
<link rel="preload" href="/assets/js/main.9e17d24d.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div><a href="#main" class="skipToContent_1oUP">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle" type="button" tabindex="0"><svg aria-label="Menu" width="30" height="30" viewBox="0 0 30 30" role="img" focusable="false"><title>Menu</title><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/"><img src="/img/logo/apple-icon-114x114.png" alt="corde logo" class="themedImage_1VuW themedImage--light_3UqQ navbar__logo"><img src="/img/logo/apple-icon-114x114.png" alt="corde logo" class="themedImage_1VuW themedImage--dark_hz6m navbar__logo"><strong class="navbar__title">Corde</strong></a><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/docs/">Documentation</a><a class="navbar__item navbar__link" href="/docs/structures/">Structures</a><a class="navbar__item navbar__link" href="/blog">Blog</a></div><div class="navbar__items navbar__items--right"><div class="navbar__item dropdown dropdown--hoverable dropdown--right"><a class="navbar__item navbar__link" href="/docs/1.x/">1.x</a><ul class="dropdown__menu"><li><a class="dropdown__link" href="/docs/">4.x</a></li><li><a class="dropdown__link" href="/docs/3.x/">3.x</a></li><li><a class="dropdown__link" href="/docs/2.x/">2.x</a></li><li><a aria-current="page" class="dropdown__link dropdown__link--active" href="/docs/1.x/command">1.x</a></li><li><a class="dropdown__link" href="/versions">All versions</a></li></ul></div><a href="https://github.com/cordejs/corde" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="react-toggle displayOnlyInLargeViewport_GrZ2 react-toggle--disabled" role="button" tabindex="-1"><div class="react-toggle-track"><div class="react-toggle-track-check"><span class="toggle_71bT">🌜</span></div><div class="react-toggle-track-x"><span class="toggle_71bT">🌞</span></div></div><div class="react-toggle-thumb"></div><input type="checkbox" class="react-toggle-screenreader-only" aria-label="Switch between dark and light mode"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div><div class="navbar-sidebar"><div class="navbar-sidebar__brand"><a class="navbar__brand" href="/"><img src="/img/logo/apple-icon-114x114.png" alt="corde logo" class="themedImage_1VuW themedImage--light_3UqQ navbar__logo"><img src="/img/logo/apple-icon-114x114.png" alt="corde logo" class="themedImage_1VuW themedImage--dark_hz6m navbar__logo"><strong class="navbar__title">Corde</strong></a></div><div class="navbar-sidebar__items"><div class="menu"><ul class="menu__list"><li class="menu__list-item"><a aria-current="page" class="menu__link navbar__link--active" href="/docs/">Documentation</a></li><li class="menu__list-item"><a class="menu__link" href="/docs/structures/">Structures</a></li><li class="menu__list-item"><a class="menu__link" href="/blog">Blog</a></li><li class="menu__list-item"><a role="button" class="menu__link menu__link--sublist">Versions</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" href="/docs/">4.x</a></li><li class="menu__list-item"><a class="menu__link" href="/docs/3.x/">3.x</a></li><li class="menu__list-item"><a class="menu__link" href="/docs/2.x/">2.x</a></li><li class="menu__list-item"><a aria-current="page" class="menu__link menu__link--active" href="/docs/1.x/command">1.x</a></li><li class="menu__list-item"><a class="menu__link" href="/versions">All versions</a></li></ul></li><li class="menu__list-item"><a href="https://github.com/cordejs/corde" target="_blank" rel="noopener noreferrer" class="menu__link header-github-link" aria-label="GitHub repository"></a></li></ul></div></div></div></nav><div class="main-wrapper docs-wrapper doc-page"><div class="docPage_31aa"><div class="docSidebarContainer_3Kbt" role="complementary"><div class="sidebar_15mo"><div class="menu menu--responsive thin-scrollbar menu_Bmed"><button aria-label="Open menu" aria-haspopup="true" class="button button--secondary button--sm menu__button" type="button"><svg aria-label="Menu" class="sidebarMenuIcon_fgN0" width="24" height="24" viewBox="0 0 30 30" role="img" focusable="false"><title>Menu</title><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><ul class="menu__list"><li class="menu__list-item"><a class="menu__link menu__link--sublist" href="#!">Introduction</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/docs/1.x/">Getting Started</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/docs/1.x/cli">CLI</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/docs/1.x/configurations">Configuring Corde</a></li></ul></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#!">API Reference</a><ul class="menu__list"><li class="menu__list-item"><a aria-current="page" class="menu__link menu__link--active active" tabindex="0" href="/docs/1.x/command">Command</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/docs/1.x/hooks">Hooks</a></li></ul></li></ul></div></div></div><main class="docMainContainer_3ufF"><div class="container padding-vert--lg docItemWrapper_3FMP"><div class="row"><div class="col docItemCol_3FnS"><div class="alert alert--warning margin-bottom--md" role="alert"><div>This is documentation for Corde <strong>1.x</strong>, which is no longer actively maintained.</div><div class="margin-top--md">For up-to-date documentation, see the <strong><a href="/docs/">latest version</a></strong> (4.x).</div></div><div class="docItemContainer_33ec"><article><div><span class="badge badge--secondary">Version: 1.x</span></div><header><h1 class="docTitle_3a4h">Command</h1></header><div class="markdown"><p>When testing functionalities with an bot, <code>expect</code> will let you define what a bot should do
given a command passed to it. All tests current available in this version(2.0) are listed bellow:</p><ul><li><a href="/docs/expect#expectcommandname-string">command(value)</a></li><li><a href="/docs/expect#not">not</a></li><li><a href="/docs/expect#toaddreactionreaction-string">mustAddReaction(value)</a></li><li><a href="/docs/expect#toreturn">mustReturn(value)</a></li></ul><div class="admonition admonition-note alert alert--secondary"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="14" height="16" viewBox="0 0 14 16"><path fill-rule="evenodd" d="M6.3 5.69a.942.942 0 0 1-.28-.7c0-.28.09-.52.28-.7.19-.18.42-.28.7-.28.28 0 .52.09.7.28.18.19.28.42.28.7 0 .28-.09.52-.28.7a1 1 0 0 1-.7.3c-.28 0-.52-.11-.7-.3zM8 7.99c-.02-.25-.11-.48-.31-.69-.2-.19-.42-.3-.69-.31H6c-.27.02-.48.13-.69.31-.2.2-.3.44-.31.69h1v3c.02.27.11.5.31.69.2.2.42.31.69.31h1c.27 0 .48-.11.69-.31.2-.19.3-.42.31-.69H8V7.98v.01zM7 2.3c-3.14 0-5.7 2.54-5.7 5.68 0 3.14 2.56 5.7 5.7 5.7s5.7-2.55 5.7-5.7c0-3.15-2.56-5.69-5.7-5.69v.01zM7 .98c3.86 0 7 3.14 7 7s-3.14 7-7 7-7-3.12-7-7 3.14-7 7-7z"></path></svg></span>note</h5></div><div class="admonition-content"><p>All examples bellow are only to ilustrate the functions called after
<code>command</code>. These examples are not used exatly show they are described
in a real use case.</p></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2LWZ" id="reference"></a>Reference<a class="hash-link" href="#reference" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2LWZ" id="mustreturnstring"></a>mustReturn(string)<a class="hash-link" href="#mustreturnstring" title="Direct link to heading">#</a></h3><p>Expect is a function that receives a command as argument. The same argument that you would send to
Discord to invoke some action from your bot. As there is a <a href="/docs/configurations#botprefix">prefix</a> witch can
be passed in config file, there is no need to pass that prefix in <code>command</code> function.</p><p>I.E:</p><p>Suposing that there is a bot with prefix <strong>!</strong> and that bot has a command called <strong>createAGuild</strong>, you can
use <code>command</code> like the example bellow:</p><div class="codeBlockContainer_K1bP"><div class="codeBlockContent_hGly javascript"><div tabindex="0" class="prism-code language-javascript codeBlock_23N8 thin-scrollbar"><div class="codeBlockLines_39YC" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token function" style="color:rgb(130, 170, 255)">command</span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token string" style="color:rgb(195, 232, 141)">&quot;createAGuild&quot;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token punctuation" style="color:rgb(199, 146, 234)">;</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_Ue-o">Copy</button></div></div><p>This function returns a several other functions to define what Corde should check after that command be invoke.</p><div class="admonition admonition-note alert alert--secondary"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="14" height="16" viewBox="0 0 14 16"><path fill-rule="evenodd" d="M6.3 5.69a.942.942 0 0 1-.28-.7c0-.28.09-.52.28-.7.19-.18.42-.28.7-.28.28 0 .52.09.7.28.18.19.28.42.28.7 0 .28-.09.52-.28.7a1 1 0 0 1-.7.3c-.28 0-.52-.11-.7-.3zM8 7.99c-.02-.25-.11-.48-.31-.69-.2-.19-.42-.3-.69-.31H6c-.27.02-.48.13-.69.31-.2.2-.3.44-.31.69h1v3c.02.27.11.5.31.69.2.2.42.31.69.31h1c.27 0 .48-.11.69-.31.2-.19.3-.42.31-.69H8V7.98v.01zM7 2.3c-3.14 0-5.7 2.54-5.7 5.68 0 3.14 2.56 5.7 5.7 5.7s5.7-2.55 5.7-5.7c0-3.15-2.56-5.69-5.7-5.69v.01zM7 .98c3.86 0 7 3.14 7 7s-3.14 7-7 7-7-3.12-7-7 3.14-7 7-7z"></path></svg></span>note</h5></div><div class="admonition-content"><p>If for some reason you what to pass the bot prefix in <code>command</code>, do not use the config variable <a href="/docs/1.x/configurations#botprefix">prefix</a>.</p></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2LWZ" id="not"></a>not<a class="hash-link" href="#not" title="Direct link to heading">#</a></h3><p>Use <code>not</code> to deny some test after this keyword. It can be used to check if a operation
made by a bot command is anything but <strong>not</strong> something.</p><p>I.E:</p><div class="codeBlockContainer_K1bP"><div class="codeBlockContent_hGly javascript"><div tabindex="0" class="prism-code language-javascript codeBlock_23N8 thin-scrollbar"><div class="codeBlockLines_39YC" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token function" style="color:rgb(130, 170, 255)">test</span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token string" style="color:rgb(195, 232, 141)">&quot;command hello should not return 1&quot;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">,</span><span class="token plain"> </span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token plain"> </span><span class="token arrow operator" style="color:rgb(137, 221, 255)">=&gt;</span><span class="token plain"> </span><span class="token punctuation" style="color:rgb(199, 146, 234)">{</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">  </span><span class="token function" style="color:rgb(130, 170, 255)">command</span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token string" style="color:rgb(195, 232, 141)">&quot;hello&quot;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token property-access">not</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token method function property-access" style="color:rgb(130, 170, 255)">mustReturn</span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token punctuation" style="color:rgb(199, 146, 234)">;</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token punctuation" style="color:rgb(199, 146, 234)">}</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token punctuation" style="color:rgb(199, 146, 234)">;</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_Ue-o">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2LWZ" id="mustaddreactionstring--string"></a>mustAddReaction(string | string[])<a class="hash-link" href="#mustaddreactionstring--string" title="Direct link to heading">#</a></h3><p>Use <code>mustAddReaction</code> to check if the bot added an reaction or an list of reactions to <strong>the last message sent by someone who is not a bot</strong>.</p><p>I.E:</p><p>Given a bot who has a command called <code>addReactions</code> and add the reactions 😄 and 🍊,
you can write a test like:</p><div class="codeBlockContainer_K1bP"><div class="codeBlockContent_hGly typescript"><div tabindex="0" class="prism-code language-typescript codeBlock_23N8 thin-scrollbar"><div class="codeBlockLines_39YC" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token function" style="color:rgb(130, 170, 255)">test</span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token string" style="color:rgb(195, 232, 141)">&quot;bot should return Hello!!&quot;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">,</span><span class="token plain"> </span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token plain"> </span><span class="token arrow operator" style="color:rgb(137, 221, 255)">=&gt;</span><span class="token plain"> </span><span class="token punctuation" style="color:rgb(199, 146, 234)">{</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">  </span><span class="token function" style="color:rgb(130, 170, 255)">command</span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token string" style="color:rgb(195, 232, 141)">&quot;addReactions&quot;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token punctuation" style="color:rgb(199, 146, 234)">.</span><span class="token method function property-access" style="color:rgb(130, 170, 255)">mustAddReaction</span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token string" style="color:rgb(195, 232, 141)">&quot;😄&quot;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">,</span><span class="token plain"> </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;🍊&quot;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token punctuation" style="color:rgb(199, 146, 234)">;</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token punctuation" style="color:rgb(199, 146, 234)">}</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token punctuation" style="color:rgb(199, 146, 234)">;</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_Ue-o">Copy</button></div></div></div></article><div class="margin-vert--xl"><div class="row"><div class="col"><a href="https://github.com/cordejs/corde/blob/master/website/versioned_docs/version-1.0.0/expect.md" target="_blank" rel="noreferrer noopener"><svg fill="currentColor" height="1.2em" width="1.2em" preserveAspectRatio="xMidYMid meet" role="img" viewBox="0 0 40 40" class="iconEdit_2_ui" aria-label="Edit page"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col text--right"><em><small>Last updated on <time datetime="2021-05-14T22:20:10.000Z" class="lastUpdatedDate_1WI_">5/14/2021</time> by <strong>Rander Oliveira</strong></small></em></div></div></div><div class="margin-vert--lg"><nav class="pagination-nav" aria-label="Docs pages navigation"><div class="pagination-nav__item"><a class="pagination-nav__link" href="/docs/1.x/configurations"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">« Configuring Corde</div></a></div><div class="pagination-nav__item pagination-nav__item--next"><a class="pagination-nav__link" href="/docs/1.x/hooks"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Hooks »</div></a></div></nav></div></div></div><div class="col col--3"><div class="tableOfContents_35-E thin-scrollbar"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#reference" class="table-of-contents__link">Reference</a><ul><li><a href="#mustreturnstring" class="table-of-contents__link">mustReturn(string)</a></li><li><a href="#not" class="table-of-contents__link">not</a></li><li><a href="#mustaddreactionstring--string" class="table-of-contents__link">mustAddReaction(string | string[])</a></li></ul></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><h4 class="footer__title">Docs</h4><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/docs/">Style Guide</a></li></ul></div><div class="col footer__col"><h4 class="footer__title">Community</h4><ul class="footer__items"><li class="footer__item"><a href="https://stackoverflow.com/questions/tagged/corde" target="_blank" rel="noopener noreferrer" class="footer__link-item">Stack Overflow</a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Corde 2021 </div></div></div></footer></div>
<script src="/assets/js/runtime~main.de5db1e2.js"></script>
<script src="/assets/js/main.9e17d24d.js"></script>
</body>
</html>