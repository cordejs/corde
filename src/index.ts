export { group, test, group as describe, test as it } from "./closures";
export { expect } from "./expect";
export { afterAll, beforeStart, afterEach, beforeEach } from "./hooks";
export * from "./types/general";

import { corde } from "./corde";
export const bot = corde.bot;

export default corde;
