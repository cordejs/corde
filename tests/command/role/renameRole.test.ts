import MockDiscord from "../../mocks/mockDiscord";
import { createCordeBotWithMockedFunctions, testHelper } from "../../testHelper";
import { ICordeBot, ITestReport } from "../../../src/types";
import { MockEvents } from "../../mocks/mockEvents";
import { debugCommand } from "../../../src/command";

const testName = "renameRole";

const failReport: ITestReport = {
  pass: false,
  testName,
};

const passReport: ITestReport = {
  pass: true,
  testName,
};

let mockDiscord = new MockDiscord();
let cordeClient = createCordeBotWithMockedFunctions(mockDiscord, mockDiscord.client);

function debugCon(customCommand?: string, customChannelId?: string, customClient?: ICordeBot) {
  return debugCommand(customCommand ?? "con", customChannelId, customClient ?? cordeClient);
}

describe(`testing ${testName} function`, () => {
  beforeEach(() => {
    [mockDiscord, cordeClient] = testHelper.initCommandTestsFixtures();
  });

  it("should fail due to undefined roleIdentifier", async () => {
    const report = await debugCon().should.renameRole(
      "egg",
      // @ts-expect-error
      "",
    );

    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should return false due to invalid newName value (null)", async () => {
    // @ts-ignore
    const report = await debugCon().should.renameRole({}, { id: "123" });
    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should return false due to newName be a empty string", async () => {
    const report = await debugCon().should.renameRole("", { id: "123" });
    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should return false due to not found role", async () => {
    cordeClient.findRole = jest.fn().mockReturnValue(null);
    const report = await debugCon().should.renameRole("newName", { id: "123" });
    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to no role was renamed", async () => {
    const report = await debugCon().should.renameRole("newName", { id: "123" });
    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should return true due to isNot true and no role change", async () => {
    const report = await debugCon().should.not.renameRole("newName", { id: "123" });
    expect(report).toEqual(passReport);
    expect(report).toMatchSnapshot();
  });

  it("should return true due to role has changed the name (isNot false)", async () => {
    const mockEvent = new MockEvents(cordeClient, mockDiscord);
    mockEvent.mockOnceRoleRenamed(mockDiscord.role);
    const report = await debugCon().should.renameRole(mockDiscord.role.name, { id: "123" });

    expect(report).toEqual(passReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a failed test due to name should not change (isNot true)", async () => {
    const mockEvent = new MockEvents(cordeClient, mockDiscord);
    mockEvent.mockOnceRoleRenamed(mockDiscord.role);
    const report = await debugCon().should.not.renameRole(mockDiscord.role.name, { id: "123" });

    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a failed test due expected name did not match to received", async () => {
    const mockEvent = new MockEvents(cordeClient, mockDiscord);
    mockEvent.mockOnceRoleRenamed(mockDiscord.role);

    const report = await debugCon().should.renameRole("test", { id: "123" });

    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a failed test due to failure in message sending", async () => {
    cordeClient.findRole = jest.fn().mockReturnValue(mockDiscord.role);
    cordeClient.fetchRole = jest.fn().mockReturnValue(null);

    const erroMessage = "can not send message to channel x";
    cordeClient.sendTextMessage = jest
      .fn()
      .mockImplementation(() => Promise.reject(new Error(erroMessage)));

    const report = await debugCon().should.renameRole("test", { id: "123" });

    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });
});
