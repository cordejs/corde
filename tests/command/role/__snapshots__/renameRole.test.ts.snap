// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing renameRole function should fail due to no role was renamed 1`] = `
Object {
  "message": "    
    expected: role 'undefined' to be renamed to newName
    received: name was not changed
",
  "pass": false,
  "testName": "renameRole",
}
`;

exports[`testing renameRole function should fail due to undefined roleIdentifier 1`] = `
Object {
  "message": "    
    expected: role 'undefined' to be renamed to egg
    received: name was not changed
",
  "pass": false,
  "testName": "renameRole",
}
`;

exports[`testing renameRole function should return a failed test due expected name did not match to received 1`] = `
Object {
  "message": "    
    expected: role change name to 'test'
    received: name was not changed (actual: 'WE DEM BOYZZ!!!!!! 1')
",
  "pass": false,
  "testName": "renameRole",
}
`;

exports[`testing renameRole function should return a failed test due to failure in message sending 1`] = `
Object {
  "message": "    
    can not send message to channel x
",
  "pass": false,
  "testName": "renameRole",
}
`;

exports[`testing renameRole function should return a failed test due to name should not change (isNot true) 1`] = `
Object {
  "message": "    
    expected: role not change name to 'WE DEM BOYZZ!!!!!! 1'
    received: name was not changed (actual: 'WE DEM BOYZZ!!!!!! 1')
",
  "pass": false,
  "testName": "renameRole",
}
`;

exports[`testing renameRole function should return false due to invalid newName value (null) 1`] = `
Object {
  "message": "    
    expected: parameter newName must be a string or a number
    received: object
",
  "pass": false,
  "testName": "renameRole",
}
`;

exports[`testing renameRole function should return false due to newName be a empty string 1`] = `
Object {
  "message": "    
    expected: parameter newName must be a valid string
    received: ''
",
  "pass": false,
  "testName": "renameRole",
}
`;

exports[`testing renameRole function should return false due to not found role 1`] = `
Object {
  "message": "    
    expected: role with id 123
    received: null
",
  "pass": false,
  "testName": "renameRole",
}
`;

exports[`testing renameRole function should return true due to isNot true and no role change 1`] = `
Object {
  "pass": true,
  "testName": "renameRole",
}
`;

exports[`testing renameRole function should return true due to role has changed the name (isNot false) 1`] = `
Object {
  "pass": true,
  "testName": "renameRole",
}
`;
